# Data Cleaning With MySQL

## Project by Benard Mwinzi


## Project Objectives
The purpose of this project is to use MySQL in jupyter notebook to clean the Nashville Housing Data. The project objectives shall be divided into;

i) Standardizing the date format.

ii) Populating property address data.

iii) Breaking out the Address column into individual columns (Address, City, State).

iv) Change Y and N into Yes and No in the "SoldAsVacant" column.

v) Deleting irrelevant columns.

vi) Create a SaleMonth column.

## Importing all necessary libraries

The libraries will enable us to use MySQL on jupyter notebook

In [2]:
#Loading necessary libraries
import pymysql
import pandas as pd

## Connecting to MySQL Database

In [3]:
# % -- 
%load_ext sql

In [4]:
#connecting to the database
%sql mysql+mysqldb://root:Your_Password@localhost/housing_data

## Creating an sql table

In [7]:
%%sql
DROP TABLE IF EXISTS housing_data;

 * mysql+mysqldb://root:***@localhost/housing_data
0 rows affected.


[]

In [8]:
%%sql
CREATE TABLE housing_data (
    UniqueID INT NOT NULL,
    ParcelID VARCHAR(50),
    LandUse VARCHAR(50),
    PropertyAddress VARCHAR(70),
    SaleDate DATE,
    SalePrice DECIMAL(15,2),
    LegalReference VARCHAR(50),
    SoldAsVacant VARCHAR(5),
    OwnerName VARCHAR(70),
    OwnerAddress VARCHAR (70),
    Acreage DECIMAL(8,3),
    TaxDistrict VARCHAR(30),
    LandValue DECIMAL(15,2),
    BuildingValue DECIMAL(15,2),
    TotalValue DECIMAL(15,2),
    YearBuilt INT,
    Bedrooms INT,
    FullBath INT,
    HalfBath INT,
    UNIQUE(UniqueID),
    PRIMARY KEY(UniqueID)
);

 * mysql+mysqldb://root:***@localhost/housing_data
0 rows affected.


[]

In [9]:
#checking the table
%sql DESCRIBE housing_data

 * mysql+mysqldb://root:***@localhost/housing_data
19 rows affected.


[('UniqueID', 'int', 'NO', 'PRI', None, ''),
 ('ParcelID', 'varchar(50)', 'YES', '', None, ''),
 ('LandUse', 'varchar(50)', 'YES', '', None, ''),
 ('PropertyAddress', 'varchar(70)', 'YES', '', None, ''),
 ('SaleDate', 'date', 'YES', '', None, ''),
 ('SalePrice', 'decimal(15,2)', 'YES', '', None, ''),
 ('LegalReference', 'varchar(50)', 'YES', '', None, ''),
 ('SoldAsVacant', 'varchar(5)', 'YES', '', None, ''),
 ('OwnerName', 'varchar(70)', 'YES', '', None, ''),
 ('OwnerAddress', 'varchar(70)', 'YES', '', None, ''),
 ('Acreage', 'decimal(8,3)', 'YES', '', None, ''),
 ('TaxDistrict', 'varchar(30)', 'YES', '', None, ''),
 ('LandValue', 'decimal(15,2)', 'YES', '', None, ''),
 ('BuildingValue', 'decimal(15,2)', 'YES', '', None, ''),
 ('TotalValue', 'decimal(15,2)', 'YES', '', None, ''),
 ('YearBuilt', 'int', 'YES', '', None, ''),
 ('Bedrooms', 'int', 'YES', '', None, ''),
 ('FullBath', 'int', 'YES', '', None, ''),
 ('HalfBath', 'int', 'YES', '', None, '')]

## Importing the data into the table

In [10]:
#Remove the strict sql mode for the session before importing the data to enable mysql to import black spaces in decimal data
%sql SET SESSION sql_mode = ''

 * mysql+mysqldb://root:***@localhost/housing_data
0 rows affected.


[]

Loading the data

In [11]:
%%sql
LOAD DATA INFILE 'D:/Data Science/MySQL/MySQL Data Cleaning/data_cleaning/Scripts/Nashville Housing Data for Data Cleaning.csv'
INTO TABLE housing_data
FIELDS TERMINATED BY ',' ENCLOSED BY '"'
LINES TERMINATED BY '\n'
IGNORE 1 LINES;

 * mysql+mysqldb://root:***@localhost/housing_data
56477 rows affected.


[]

## Cleaning the data

Checking the first few rows

In [12]:
%%sql 
SELECT * FROM housing_data
LIMIT 5;

 * mysql+mysqldb://root:***@localhost/housing_data
5 rows affected.


[(0, '105 03 0D 008.00', 'RESIDENTIAL CONDO', '1208  3RD AVE S, NASHVILLE', datetime.date(2013, 1, 24), Decimal('132000.00'), '20130128-0008725', 'No', '', '', Decimal('0.000'), '', Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), 0, 0, 0, 0),
 (1, '105 11 0 080.00', 'SINGLE FAMILY', '1802  STEWART PL, NASHVILLE', datetime.date(2013, 1, 11), Decimal('191500.00'), '20130118-0006337', 'No', 'STINSON, LAURA M.', '1802  STEWART PL, NASHVILLE, TN', Decimal('0.170'), 'URBAN SERVICES DISTRICT', Decimal('32000.00'), Decimal('134400.00'), Decimal('168300.00'), 1941, 2, 1, 0),
 (2, '118 03 0 130.00', 'SINGLE FAMILY', '2761  ROSEDALE PL, NASHVILLE', datetime.date(2013, 1, 18), Decimal('202000.00'), '20130124-0008033', 'No', 'NUNES, JARED R.', '2761  ROSEDALE PL, NASHVILLE, TN', Decimal('0.110'), 'CITY OF BERRY HILL', Decimal('34000.00'), Decimal('157800.00'), Decimal('191800.00'), 2000, 3, 2, 1),
 (3, '119 01 0 479.00', 'SINGLE FAMILY', '224  PEACHTREE ST, NASHVILLE', datetime.date(2013, 1, 18), Decimal('32000.00'), '20130128-0008863', 'No', 'WHITFORD, KAREN', '224  PEACHTREE ST, NASHVILLE, TN', Decimal('0.170'), 'URBAN SERVICES DISTRICT', Decimal('25000.00'), Decimal('243700.00'), Decimal('268700.00'), 1948, 4, 2, 0),
 (4, '119 05 0 186.00', 'SINGLE FAMILY', '316  LUTIE ST, NASHVILLE', datetime.date(2013, 1, 23), Decimal('102000.00'), '20130131-0009929', 'No', 'HENDERSON, JAMES P. & LYNN P.', '316  LUTIE ST, NASHVILLE, TN', Decimal('0.340'), 'URBAN SERVICES DISTRICT', Decimal('25000.00'), Decimal('138100.00'), Decimal('164800.00'), 1910, 2, 1, 0)]

### i) Standardizing the Date Format

Creating a converted SaleDate column

In [13]:
%%sql
SELECT SaleDate, DATE(SaleDate)
FROM housing_data
LIMIT 10;

 * mysql+mysqldb://root:***@localhost/housing_data
10 rows affected.


[(datetime.date(2013, 1, 24), datetime.date(2013, 1, 24)),
 (datetime.date(2013, 1, 11), datetime.date(2013, 1, 11)),
 (datetime.date(2013, 1, 18), datetime.date(2013, 1, 18)),
 (datetime.date(2013, 1, 18), datetime.date(2013, 1, 18)),
 (datetime.date(2013, 1, 23), datetime.date(2013, 1, 23)),
 (datetime.date(2013, 1, 4), datetime.date(2013, 1, 4)),
 (datetime.date(2013, 1, 7), datetime.date(2013, 1, 7)),
 (datetime.date(2013, 1, 15), datetime.date(2013, 1, 15)),
 (datetime.date(2013, 1, 25), datetime.date(2013, 1, 25)),
 (datetime.date(2013, 1, 9), datetime.date(2013, 1, 9))]

Updating the SaleDate column with the standardized date data

In [14]:
%%sql
UPDATE housing_data
SET SaleDate = DATE(SaleDate);

 * mysql+mysqldb://root:***@localhost/housing_data
56477 rows affected.


[]

Confirming that the date variable is standardised

In [15]:
%%sql
SELECT *
FROM housing_data
LIMIT 3;

 * mysql+mysqldb://root:***@localhost/housing_data
3 rows affected.


[(0, '105 03 0D 008.00', 'RESIDENTIAL CONDO', '1208  3RD AVE S, NASHVILLE', datetime.date(2013, 1, 24), Decimal('132000.00'), '20130128-0008725', 'No', '', '', Decimal('0.000'), '', Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), 0, 0, 0, 0),
 (1, '105 11 0 080.00', 'SINGLE FAMILY', '1802  STEWART PL, NASHVILLE', datetime.date(2013, 1, 11), Decimal('191500.00'), '20130118-0006337', 'No', 'STINSON, LAURA M.', '1802  STEWART PL, NASHVILLE, TN', Decimal('0.170'), 'URBAN SERVICES DISTRICT', Decimal('32000.00'), Decimal('134400.00'), Decimal('168300.00'), 1941, 2, 1, 0),
 (2, '118 03 0 130.00', 'SINGLE FAMILY', '2761  ROSEDALE PL, NASHVILLE', datetime.date(2013, 1, 18), Decimal('202000.00'), '20130124-0008033', 'No', 'NUNES, JARED R.', '2761  ROSEDALE PL, NASHVILLE, TN', Decimal('0.110'), 'CITY OF BERRY HILL', Decimal('34000.00'), Decimal('157800.00'), Decimal('191800.00'), 2000, 3, 2, 1)]

### ii) Populating the property address data because it has some missing values

Checking the missing values

In [16]:
%%sql
SELECT *
FROM housing_data
WHERE PropertyAddress = "";

 * mysql+mysqldb://root:***@localhost/housing_data
29 rows affected.


[(3299, '052 01 0 296.00', 'SINGLE FAMILY', '', datetime.date(2013, 5, 31), Decimal('79370.00'), '20130620-0063114', 'No', 'KIRBY, WILLIAM E. JR', '726  IDLEWILD DR, MADISON, TN', Decimal('0.220'), 'GENERAL SERVICES DISTRICT', Decimal('16000.00'), Decimal('60800.00'), Decimal('76800.00'), 1957, 2, 1, 0),
 (8126, '093 08 0 054.00', 'SINGLE FAMILY', '', datetime.date(2013, 9, 20), Decimal('25000.00'), '20130923-0099666', 'N', 'CRAWFORD, CHRISTOPHER C. & BOBBIE LEANN', '700  GLENVIEW DR, NASHVILLE, TN', Decimal('0.210'), 'URBAN SERVICES DISTRICT', Decimal('25000.00'), Decimal('236800.00'), Decimal('261800.00'), 2015, 3, 2, 1),
 (11478, '092 13 0 322.00', 'SINGLE FAMILY', '', datetime.date(2014, 1, 17), Decimal('269500.00'), '20140122-0006168', 'No', 'WIGGINS, MATHEW L. & GOOD, KENDRA A.', '237  37TH AVE N, NASHVILLE, TN', Decimal('0.340'), 'URBAN SERVICES DISTRICT', Decimal('91000.00'), Decimal('107100.00'), Decimal('198100.00'), 1945, 2, 1, 0),
 (14753, '108 07 0A 026.00', 'RESIDENTIAL CONDO', '', datetime.date(2014, 4, 15), Decimal('79900.00'), '20140416-0031777', 'No', '', '', Decimal('0.000'), '', Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), 0, 0, 0, 0),
 (15886, '109 04 0A 080.00', 'VACANT RES LAND', '', datetime.date(2014, 5, 13), Decimal('255590.00'), '20140514-0041276', 'No', '', '', Decimal('0.000'), '', Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), 0, 0, 0, 0),
 (22775, '043 09 0 074.00', 'VACANT RESIDENTIAL LAND', '', datetime.date(2014, 10, 27), Decimal('151000.00'), '20141028-0099094', 'Yes', 'BREWER HOLDINGS, LLC', '213 B LOVELL ST, MADISON, TN', Decimal('0.150'), 'GENERAL SERVICES DISTRICT', Decimal('18000.00'), Decimal('115600.00'), Decimal('133600.00'), 2015, 3, 2, 0),
 (24197, '110 03 0A 061.00', 'SINGLE FAMILY', '', datetime.date(2014, 11, 19), Decimal('269750.00'), '20141120-0106962', 'No', '', '', Decimal('0.000'), '', Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), 0, 0, 0, 0),
 (27140, '092 06 0 282.00', 'SINGLE FAMILY', '', datetime.date(2015, 2, 20), Decimal('41500.00'), '20150224-0015900', 'No', 'PROSPERITAS PARTNERS LLC', '815  31ST AVE N, NASHVILLE, TN', Decimal('0.170'), 'URBAN SERVICES DISTRICT', Decimal('13000.00'), Decimal('49400.00'), Decimal('62400.00'), 1960, 2, 1, 0),
 (32385, '092 13 0 339.00', 'SINGLE FAMILY', '', datetime.date(2015, 6, 5), Decimal('450000.00'), '20150618-0058311', 'No', 'JUMPER, NICOLE', '311  35TH AVE N, NASHVILLE, TN', Decimal('0.160'), 'URBAN SERVICES DISTRICT', Decimal('65000.00'), Decimal('365300.00'), Decimal('430300.00'), 2015, 3, 3, 1),
 (36531, '034 03 0 059.00', 'SINGLE FAMILY', '', datetime.date(2015, 8, 13), Decimal('245000.00'), '20150819-0083759', 'No', 'DILICK, JOHN MARK & ANNETTE A.', '2117  PAULA DR, MADISON, TN', Decimal('1.010'), 'GENERAL SERVICES DISTRICT', Decimal('32000.00'), Decimal('170000.00'), Decimal('228300.00'), 1964, 4, 3, 0),
 (39432, '026 01 0 069.00', 'VACANT RESIDENTIAL LAND', '', datetime.date(2015, 10, 23), Decimal('153000.00'), '20151028-0109602', 'No', 'SHACKLEFORD, MICHAEL C., JR.', '141  TWO MILE PIKE, GOODLETTSVILLE, TN', Decimal('0.170'), 'CITY OF GOODLETTSVILLE', Decimal('21100.00'), Decimal('121600.00'), Decimal('142700.00'), 2015, 3, 2, 0),
 (40678, '042 13 0 075.00', 'SINGLE FAMILY', '', datetime.date(2015, 11, 30), Decimal('208000.00'), '20151209-0123831', 'No', 'SADOWSKY, ERIC C. & LISA D.', '222  FOXBORO DR, MADISON, TN', Decimal('1.010'), 'GENERAL SERVICES DISTRICT', Decimal('30000.00'), Decimal('100000.00'), Decimal('130000.00'), 1967, 3, 1, 1),
 (43076, '025 07 0 031.00', 'SINGLE FAMILY', '', datetime.date(2016, 1, 15), Decimal('179900.00'), '20160120-0005776', 'No', 'COSTNER, FRED & CAROLYN', '410  ROSEHILL CT, GOODLETTSVILLE, TN', Decimal('0.960'), 'CITY OF GOODLETTSVILLE', Decimal('30000.00'), Decimal('70000.00'), Decimal('100000.00'), 1964, 3, 1, 0),
 (43080, '033 06 0 041.00', 'SINGLE FAMILY', '', datetime.date(2016, 1, 4), Decimal('170000.00'), '20160107-0001526', 'No', 'FRANK, ZACHARY & NIKI', '1129  CA

As seen in the above output, there are 29 rows with missing values on the PropertyAddress variable.

The missing points in the property address were populated by referencing rows with properties that are in the same parcel.

Populating the PropertyAddress column data is possible because each property belongs to a parcel. Therefore, houses in the same parcel share the same PropertyAddress and percel. We are going to use the addresses of the other properties in the same percelIds to populate the data.



Updating the table with the new values to replace the missing values in PropertyAddress

In [17]:
%%sql
SELECT a.UniqueID, b.ParcelID, b.PropertyAddress, IF (a.PropertyAddress = "", b.PropertyAddress, a.PropertyAddress) updated
FROM housing_data a, housing_data b
WHERE a.ParcelID = b.ParcelID
AND a.UniqueID <> b.UniqueID
AND a.PropertyAddress = "";


 * mysql+mysqldb://root:***@localhost/housing_data
35 rows affected.


[(11478, '092 13 0 322.00', '237  37TH AVE N, NASHVILLE', '237  37TH AVE N, NASHVILLE'),
 (47293, '043 04 0 014.00', '112  HILLER DR, OLD HICKORY', '112  HILLER DR, OLD HICKORY'),
 (45290, '026 05 0 017.00', '208  EAST AVE, GOODLETTSVILLE', '208  EAST AVE, GOODLETTSVILLE'),
 (40678, '042 13 0 075.00', '222  FOXBORO DR, MADISON', '222  FOXBORO DR, MADISON'),
 (43151, '052 08 0A 320.00', '608  SANDY SPRING TRL, MADISON', '608  SANDY SPRING TRL, MADISON'),
 (45774, '107 13 0 107.00', '1205  THOMPSON PL, NASHVILLE', '1205  THOMPSON PL, NASHVILLE'),
 (43080, '033 06 0 041.00', '1129  CAMPBELL RD, GOODLETTSVILLE', '1129  CAMPBELL RD, GOODLETTSVILLE'),
 (8126, '093 08 0 054.00', '700  GLENVIEW DR, NASHVILLE', '700  GLENVIEW DR, NASHVILLE'),
 (45295, '033 06 0A 002.00', '1116  CAMPBELL RD, GOODLETTSVILLE', '1116  CAMPBELL RD, GOODLETTSVILLE'),
 (15886, '109 04 0A 080.00', '2537  JANALYN TRCE, HERMITAGE', '2537  JANALYN TRCE, HERMITAGE'),
 (24197, '110 03 0A 061.00', '2704  ALVIN SPERRY PASS, MOUNT JULIET', '2704  ALVIN SPERRY PASS, MOUNT JULIET'),
 (53147, '026 06 0A 038.00', '109  CANTON CT, GOODLETTSVILLE', '109  CANTON CT, GOODLETTSVILLE'),
 (32385, '092 13 0 339.00', '311  35TH AVE N, NASHVILLE', '311  35TH AVE N, NASHVILLE'),
 (39432, '026 01 0 069.00', '141  TWO MILE PIKE, GOODLETTSVILLE', '141  TWO MILE PIKE, GOODLETTSVILLE'),
 (11478, '092 13 0 322.00', '237  37TH AVE N, NASHVILLE', '237  37TH AVE N, NASHVILLE'),
 (46919, '034 07 0B 015.00', '2524  VAL MARIE DR, MADISON', '2524  VAL MARIE DR, MADISON'),
 (27140, '092 06 0 282.00', '815  31ST AVE N, NASHVILLE', '815  31ST AVE N, NASHVILLE'),
 (22775, '043 09 0 074.00', '213 B  LOVELL ST, MADISON', '213 B  LOVELL ST, MADISON'),
 (45298, '041 03 0A 100.00', '1289  GOODMORNING DR, NASHVILLE', '1289  GOODMORNING DR, NASHVILLE'),
 (51930, '113 14 0A 002.00', '7601  CHIPMUNK LN, NASHVILLE', '7601  CHIPMUNK LN, NASHVILLE'),
 (36531, '034 03 0 059.00', '2117  PAULA DR, MADISON', '2117  PAULA DR, MADISON'),
 (51703, '114 15 0A 030.00', '109  CEDAR PLACE BND, NASHVILLE', '109  CEDAR PLACE BND, NASHVILLE'),
 (36531, '034 03 0 059.00', '2117  PAULA DR, MADISON', '2117  PAULA DR, MADISON'),
 (8126, '093 08 0 054.00', '700  GLENVIEW DR, NASHVILLE', '700  GLENVIEW DR, NASHVILLE'),
 (43076, '025 07 0 031.00', '410  ROSEHILL CT, GOODLETTSVILLE', '410  ROSEHILL CT, GOODLETTSVILLE'),
 (3299, '052 01 0 296.00', '726  IDLEWILD DR, MADISON', '726  IDLEWILD DR, MADISON'),
 (50927, '044 05 0 135.00', '202  KEETON AVE, OLD HICKORY', '202  KEETON AVE, OLD HICKORY'),
 (14753, '108 07 0A 026.00', '908  PATIO DR, NASHVILLE', '908  PATIO DR, NASHVILLE'),
 (48731, '033 15 0 123.00', '438  W CAMPBELL RD, GOODLETTSVILLE', '438  W CAMPBELL RD, GOODLETTSVILLE'),
 (45349, '043 13 0 308.00', '224  HICKORY ST, MADISON', '224  HICKORY ST, MADISON'),
 (49886, '092 06 0 273.00', '2721  HERMAN ST, NASHVILLE', '2721  HERMAN ST, NASHVILLE'),
 (14753, '108 07 0A 026.00', '908  PATIO DR, NASHVILLE', '908  PATIO DR, NASHVILLE'),
 (46919, '034 07 0B 015.00', '2524  VAL MARIE DR, MADISON', '2524  VAL MARIE DR, MADISON'),
 (46919, '034 07 0B 015.00', '2524 VAL MARIE  DR, MADISON', '2524 VAL MARIE  DR, MADISON'),
 (44264, '034 16 0A 004.00', '213 WARREN  CT, OLD HICKORY', '213 WARREN  CT, OLD HICKORY')]

The "updated" column in the table above can populate the missing values in our data. We are going to first save the above output in a view to make it easy to populate our data.

In [18]:
%%sql
CREATE VIEW updated_address AS
SELECT a.UniqueID, b.ParcelID, b.PropertyAddress, IF (a.PropertyAddress = "", b.PropertyAddress, a.PropertyAddress) updated_address
FROM housing_data a, housing_data b
WHERE a.ParcelID = b.ParcelID
AND a.UniqueID <> b.UniqueID
AND a.PropertyAddress = "";


 * mysql+mysqldb://root:***@localhost/housing_data
0 rows affected.


[]

We can check our updated_address view

In [19]:
%sql select * from updated_address;

 * mysql+mysqldb://root:***@localhost/housing_data
35 rows affected.


[(11478, '092 13 0 322.00', '237  37TH AVE N, NASHVILLE', '237  37TH AVE N, NASHVILLE'),
 (47293, '043 04 0 014.00', '112  HILLER DR, OLD HICKORY', '112  HILLER DR, OLD HICKORY'),
 (45290, '026 05 0 017.00', '208  EAST AVE, GOODLETTSVILLE', '208  EAST AVE, GOODLETTSVILLE'),
 (40678, '042 13 0 075.00', '222  FOXBORO DR, MADISON', '222  FOXBORO DR, MADISON'),
 (43151, '052 08 0A 320.00', '608  SANDY SPRING TRL, MADISON', '608  SANDY SPRING TRL, MADISON'),
 (45774, '107 13 0 107.00', '1205  THOMPSON PL, NASHVILLE', '1205  THOMPSON PL, NASHVILLE'),
 (43080, '033 06 0 041.00', '1129  CAMPBELL RD, GOODLETTSVILLE', '1129  CAMPBELL RD, GOODLETTSVILLE'),
 (8126, '093 08 0 054.00', '700  GLENVIEW DR, NASHVILLE', '700  GLENVIEW DR, NASHVILLE'),
 (45295, '033 06 0A 002.00', '1116  CAMPBELL RD, GOODLETTSVILLE', '1116  CAMPBELL RD, GOODLETTSVILLE'),
 (15886, '109 04 0A 080.00', '2537  JANALYN TRCE, HERMITAGE', '2537  JANALYN TRCE, HERMITAGE'),
 (24197, '110 03 0A 061.00', '2704  ALVIN SPERRY PASS, MOUNT JULIET', '2704  ALVIN SPERRY PASS, MOUNT JULIET'),
 (53147, '026 06 0A 038.00', '109  CANTON CT, GOODLETTSVILLE', '109  CANTON CT, GOODLETTSVILLE'),
 (32385, '092 13 0 339.00', '311  35TH AVE N, NASHVILLE', '311  35TH AVE N, NASHVILLE'),
 (39432, '026 01 0 069.00', '141  TWO MILE PIKE, GOODLETTSVILLE', '141  TWO MILE PIKE, GOODLETTSVILLE'),
 (11478, '092 13 0 322.00', '237  37TH AVE N, NASHVILLE', '237  37TH AVE N, NASHVILLE'),
 (46919, '034 07 0B 015.00', '2524  VAL MARIE DR, MADISON', '2524  VAL MARIE DR, MADISON'),
 (27140, '092 06 0 282.00', '815  31ST AVE N, NASHVILLE', '815  31ST AVE N, NASHVILLE'),
 (22775, '043 09 0 074.00', '213 B  LOVELL ST, MADISON', '213 B  LOVELL ST, MADISON'),
 (45298, '041 03 0A 100.00', '1289  GOODMORNING DR, NASHVILLE', '1289  GOODMORNING DR, NASHVILLE'),
 (51930, '113 14 0A 002.00', '7601  CHIPMUNK LN, NASHVILLE', '7601  CHIPMUNK LN, NASHVILLE'),
 (36531, '034 03 0 059.00', '2117  PAULA DR, MADISON', '2117  PAULA DR, MADISON'),
 (51703, '114 15 0A 030.00', '109  CEDAR PLACE BND, NASHVILLE', '109  CEDAR PLACE BND, NASHVILLE'),
 (36531, '034 03 0 059.00', '2117  PAULA DR, MADISON', '2117  PAULA DR, MADISON'),
 (8126, '093 08 0 054.00', '700  GLENVIEW DR, NASHVILLE', '700  GLENVIEW DR, NASHVILLE'),
 (43076, '025 07 0 031.00', '410  ROSEHILL CT, GOODLETTSVILLE', '410  ROSEHILL CT, GOODLETTSVILLE'),
 (3299, '052 01 0 296.00', '726  IDLEWILD DR, MADISON', '726  IDLEWILD DR, MADISON'),
 (50927, '044 05 0 135.00', '202  KEETON AVE, OLD HICKORY', '202  KEETON AVE, OLD HICKORY'),
 (14753, '108 07 0A 026.00', '908  PATIO DR, NASHVILLE', '908  PATIO DR, NASHVILLE'),
 (48731, '033 15 0 123.00', '438  W CAMPBELL RD, GOODLETTSVILLE', '438  W CAMPBELL RD, GOODLETTSVILLE'),
 (45349, '043 13 0 308.00', '224  HICKORY ST, MADISON', '224  HICKORY ST, MADISON'),
 (49886, '092 06 0 273.00', '2721  HERMAN ST, NASHVILLE', '2721  HERMAN ST, NASHVILLE'),
 (14753, '108 07 0A 026.00', '908  PATIO DR, NASHVILLE', '908  PATIO DR, NASHVILLE'),
 (46919, '034 07 0B 015.00', '2524  VAL MARIE DR, MADISON', '2524  VAL MARIE DR, MADISON'),
 (46919, '034 07 0B 015.00', '2524 VAL MARIE  DR, MADISON', '2524 VAL MARIE  DR, MADISON'),
 (44264, '034 16 0A 004.00', '213 WARREN  CT, OLD HICKORY', '213 WARREN  CT, OLD HICKORY')]

Using the Inner Join statement to finally populate our data.

In [20]:
%%sql
UPDATE housing_data
INNER JOIN updated_address
ON housing_data.ParcelID = updated_address.ParcelID
AND housing_data.PropertyAddress = ""
SET housing_data.PropertyAddress = updated_address.PropertyAddress;


 * mysql+mysqldb://root:***@localhost/housing_data
29 rows affected.


[]

Confirming the data is updated

In [21]:
%%sql
SELECT *
FROM housing_data
WHERE PropertyAddress = "";

 * mysql+mysqldb://root:***@localhost/housing_data
0 rows affected.


[]

There is no row with missing values in the PropertyAddress meaning that our data is properly updated.

However, we need to confirm that our data still has 56477 rows 

In [22]:
%%sql
SELECT COUNT(*)
FROM housing_data;

 * mysql+mysqldb://root:***@localhost/housing_data
1 rows affected.


[(56477,)]

The data has 56477 rows. Our PropertyAddress is now complete.

### iii) Breaking out the Address column into individual columns (Address, City, State).

Let's first check the first five rows of our data

In [23]:
%%sql
SELECT * 
FROM housing_data
LIMIT 5;

 * mysql+mysqldb://root:***@localhost/housing_data
5 rows affected.


[(0, '105 03 0D 008.00', 'RESIDENTIAL CONDO', '1208  3RD AVE S, NASHVILLE', datetime.date(2013, 1, 24), Decimal('132000.00'), '20130128-0008725', 'No', '', '', Decimal('0.000'), '', Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), 0, 0, 0, 0),
 (1, '105 11 0 080.00', 'SINGLE FAMILY', '1802  STEWART PL, NASHVILLE', datetime.date(2013, 1, 11), Decimal('191500.00'), '20130118-0006337', 'No', 'STINSON, LAURA M.', '1802  STEWART PL, NASHVILLE, TN', Decimal('0.170'), 'URBAN SERVICES DISTRICT', Decimal('32000.00'), Decimal('134400.00'), Decimal('168300.00'), 1941, 2, 1, 0),
 (2, '118 03 0 130.00', 'SINGLE FAMILY', '2761  ROSEDALE PL, NASHVILLE', datetime.date(2013, 1, 18), Decimal('202000.00'), '20130124-0008033', 'No', 'NUNES, JARED R.', '2761  ROSEDALE PL, NASHVILLE, TN', Decimal('0.110'), 'CITY OF BERRY HILL', Decimal('34000.00'), Decimal('157800.00'), Decimal('191800.00'), 2000, 3, 2, 1),
 (3, '119 01 0 479.00', 'SINGLE FAMILY', '224  PEACHTREE ST, NASHVILLE', datetime.date(2013, 1, 18), Decimal('32000.00'), '20130128-0008863', 'No', 'WHITFORD, KAREN', '224  PEACHTREE ST, NASHVILLE, TN', Decimal('0.170'), 'URBAN SERVICES DISTRICT', Decimal('25000.00'), Decimal('243700.00'), Decimal('268700.00'), 1948, 4, 2, 0),
 (4, '119 05 0 186.00', 'SINGLE FAMILY', '316  LUTIE ST, NASHVILLE', datetime.date(2013, 1, 23), Decimal('102000.00'), '20130131-0009929', 'No', 'HENDERSON, JAMES P. & LYNN P.', '316  LUTIE ST, NASHVILLE, TN', Decimal('0.340'), 'URBAN SERVICES DISTRICT', Decimal('25000.00'), Decimal('138100.00'), Decimal('164800.00'), 1910, 2, 1, 0)]

As seen in the above output, there are two address columns: PropertyAddress and OwnerAddress.

The PropertyAddress has two parts: the address and city. We deal with the PropertyAddress first, then the OwnerAddress later.

From the PropertyAddress, we shall have PropertyAddress and PropertyCity.

Writing a SELECT statement to split the variable into address and city.

In [24]:
%%sql
SELECT PropertyAddress, SUBSTRING_INDEX(PropertyAddress, ",", 1) AS PropertyAddress_update, SUBSTRING_INDEX(PropertyAddress, ",", -1) AS PropertyCity
FROM housing_data
LIMIT 5;

 * mysql+mysqldb://root:***@localhost/housing_data
5 rows affected.


[('1208  3RD AVE S, NASHVILLE', '1208  3RD AVE S', ' NASHVILLE'),
 ('1802  STEWART PL, NASHVILLE', '1802  STEWART PL', ' NASHVILLE'),
 ('2761  ROSEDALE PL, NASHVILLE', '2761  ROSEDALE PL', ' NASHVILLE'),
 ('224  PEACHTREE ST, NASHVILLE', '224  PEACHTREE ST', ' NASHVILLE'),
 ('316  LUTIE ST, NASHVILLE', '316  LUTIE ST', ' NASHVILLE')]

As seen above, we now have new variables obtained by slitting the PropertyAdress variable.

Now, we need to update our data.


We used "alter table" function to create the new variables: propertyAddressUpdate and PropertyCity

In [25]:
%%sql
ALTER TABLE housing_data
ADD PropertyAddressUpdate VARCHAR(50),
ADD PropertyCity VARCHAR(30);

 * mysql+mysqldb://root:***@localhost/housing_data
0 rows affected.


[]

populating the new variables by splitting the PropertyAddress Variable

In [27]:
%%sql
UPDATE housing_data
SET PropertyAddressUpdate = SUBSTRING_INDEX(PropertyAddress, ",", 1),
PropertyCity = TRIM(SUBSTRING_INDEX(PropertyAddress, ',', -1));

 * mysql+mysqldb://root:***@localhost/housing_data
56477 rows affected.


[]

We now drop the original PropertyAddress column from the data

In [28]:
%%sql
ALTER TABLE housing_data
DROP COLUMN PropertyAddress;

 * mysql+mysqldb://root:***@localhost/housing_data
0 rows affected.


[]

Now we can rename the PropertyAddressUpdate column to PropertyAddress

In [29]:
%%sql
ALTER TABLE housing_data
RENAME COLUMN PropertyAddressUpdate to PropertyAddress;

 * mysql+mysqldb://root:***@localhost/housing_data
0 rows affected.


[]

The PropertyAddress Column has been properly split. We now need to deal with the OwnerAddress.

Lets check the first 3 rows of the dataset

In [30]:
%%sql
SELECT * 
FROM housing_data
LIMIT 3;

 * mysql+mysqldb://root:***@localhost/housing_data
3 rows affected.


[(0, '105 03 0D 008.00', 'RESIDENTIAL CONDO', datetime.date(2013, 1, 24), Decimal('132000.00'), '20130128-0008725', 'No', '', '', Decimal('0.000'), '', Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), 0, 0, 0, 0, '1208  3RD AVE S', 'NASHVILLE'),
 (1, '105 11 0 080.00', 'SINGLE FAMILY', datetime.date(2013, 1, 11), Decimal('191500.00'), '20130118-0006337', 'No', 'STINSON, LAURA M.', '1802  STEWART PL, NASHVILLE, TN', Decimal('0.170'), 'URBAN SERVICES DISTRICT', Decimal('32000.00'), Decimal('134400.00'), Decimal('168300.00'), 1941, 2, 1, 0, '1802  STEWART PL', 'NASHVILLE'),
 (2, '118 03 0 130.00', 'SINGLE FAMILY', datetime.date(2013, 1, 18), Decimal('202000.00'), '20130124-0008033', 'No', 'NUNES, JARED R.', '2761  ROSEDALE PL, NASHVILLE, TN', Decimal('0.110'), 'CITY OF BERRY HILL', Decimal('34000.00'), Decimal('157800.00'), Decimal('191800.00'), 2000, 3, 2, 1, '2761  ROSEDALE PL', 'NASHVILLE')]

As seen in the above output, the OwnerAddress column has three parts, the address, the city and state.

We need to divide the column into three parts.

We start by writing a select statement to split the variable into OwnerAddress, city and State

In [34]:
%%sql
SELECT SUBSTRING_INDEX(OwnerAddress, ",", 1) AS OwnerAddress_update, SUBSTRING_INDEX(SUBSTRING_INDEX(OwnerAddress, ",", -2),",",1) AS City, SUBSTRING_INDEX(OwnerAddress, ",", -1) AS State
FROM housing_data
LIMIT 5;

 * mysql+mysqldb://root:***@localhost/housing_data
5 rows affected.


[('', '', ''),
 ('1802  STEWART PL', ' NASHVILLE', ' TN'),
 ('2761  ROSEDALE PL', ' NASHVILLE', ' TN'),
 ('224  PEACHTREE ST', ' NASHVILLE', ' TN'),
 ('316  LUTIE ST', ' NASHVILLE', ' TN')]

We now use the "alter table" function to create the new variables: OwnerAddress, OwnerCity and OwnerState

In [35]:
%%sql
ALTER TABLE housing_data
ADD OwnerAddressUpdate VARCHAR(50),
ADD OwnerCity VARCHAR(20),
ADD OwnerState VARCHAR(20);

 * mysql+mysqldb://root:***@localhost/housing_data
0 rows affected.


[]

Populating the new variables by splitting the PropertyAddress Variable

In [36]:
%%sql
UPDATE housing_data
SET OwnerAddressUpdate = SUBSTRING_INDEX(OwnerAddress, ",", 1),
OwnerCity = SUBSTRING_INDEX(SUBSTRING_INDEX(OwnerAddress, ",", -2),",",1),
OwnerState = SUBSTRING_INDEX(OwnerAddress, ",", -1);

 * mysql+mysqldb://root:***@localhost/housing_data
56477 rows affected.


[]

Dropping the original OwnerAddress column from the data.

In [37]:
%%sql
ALTER TABLE housing_data
DROP COLUMN OwnerAddress;

 * mysql+mysqldb://root:***@localhost/housing_data
0 rows affected.


[]

Changing the OwnerAddressUpdate column to OwnerAddress


In [38]:
%%sql
ALTER TABLE housing_data
RENAME COLUMN OwnerAddressUpdate TO OwnerAddress;

 * mysql+mysqldb://root:***@localhost/housing_data
0 rows affected.


[]

Now, the OwnwerAddress column has been split into address, city, and state.

We an check the first 5 rows to ascertain that the column was properly split.

In [39]:
%%sql
SELECT *
FROM housing_data
LIMIT 5;

 * mysql+mysqldb://root:***@localhost/housing_data
5 rows affected.


[(0, '105 03 0D 008.00', 'RESIDENTIAL CONDO', datetime.date(2013, 1, 24), Decimal('132000.00'), '20130128-0008725', 'No', '', Decimal('0.000'), '', Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), 0, 0, 0, 0, '1208  3RD AVE S', 'NASHVILLE', '', '', ''),
 (1, '105 11 0 080.00', 'SINGLE FAMILY', datetime.date(2013, 1, 11), Decimal('191500.00'), '20130118-0006337', 'No', 'STINSON, LAURA M.', Decimal('0.170'), 'URBAN SERVICES DISTRICT', Decimal('32000.00'), Decimal('134400.00'), Decimal('168300.00'), 1941, 2, 1, 0, '1802  STEWART PL', 'NASHVILLE', '1802  STEWART PL', ' NASHVILLE', ' TN'),
 (2, '118 03 0 130.00', 'SINGLE FAMILY', datetime.date(2013, 1, 18), Decimal('202000.00'), '20130124-0008033', 'No', 'NUNES, JARED R.', Decimal('0.110'), 'CITY OF BERRY HILL', Decimal('34000.00'), Decimal('157800.00'), Decimal('191800.00'), 2000, 3, 2, 1, '2761  ROSEDALE PL', 'NASHVILLE', '2761  ROSEDALE PL', ' NASHVILLE', ' TN'),
 (3, '119 01 0 479.00', 'SINGLE FAMILY', datetime.date(2013, 1, 18), Decimal('32000.00'), '20130128-0008863', 'No', 'WHITFORD, KAREN', Decimal('0.170'), 'URBAN SERVICES DISTRICT', Decimal('25000.00'), Decimal('243700.00'), Decimal('268700.00'), 1948, 4, 2, 0, '224  PEACHTREE ST', 'NASHVILLE', '224  PEACHTREE ST', ' NASHVILLE', ' TN'),
 (4, '119 05 0 186.00', 'SINGLE FAMILY', datetime.date(2013, 1, 23), Decimal('102000.00'), '20130131-0009929', 'No', 'HENDERSON, JAMES P. & LYNN P.', Decimal('0.340'), 'URBAN SERVICES DISTRICT', Decimal('25000.00'), Decimal('138100.00'), Decimal('164800.00'), 1910, 2, 1, 0, '316  LUTIE ST', 'NASHVILLE', '316  LUTIE ST', ' NASHVILLE', ' TN')]

The above output shows that the data was successfully split.

### iv) Change Y and N into Yes and No in the "SoldAsVacant" column.

Now, we need to use a CASE statement to update our table to have y where the SoldAsVacant column is "Yes" and N where the column is "No".

However, we first need to write a SELECT statement to group the column to determine if there might be a third class. may be an incorrect input.

In [40]:
%%sql
SELECT SoldAsVacant, COUNT(SoldAsVacant)
FROM housing_data
GROUP BY SoldAsVacant;

 * mysql+mysqldb://root:***@localhost/housing_data
4 rows affected.


[('No', 51403), ('N', 399), ('Y', 52), ('Yes', 4623)]

As seen in the above output, some rows already have "N" and "Y".

We are going to write a CASE statement to ensure that all the rows have either Y for Yes or N for No

In [49]:
%%sql
SELECT CASE WHEN SoldAsVacant = "Yes" OR SoldAsVacant = "Y" THEN "Y"
ELSE "N" END AS SOLD_AS_V,
COUNT(SoldAsVacant)
FROM housing_data
GROUP BY SOLD_AS_V;

 * mysql+mysqldb://root:***@localhost/housing_data
2 rows affected.


[('N', 51802), ('Y', 4675)]

From the above table, we have alredy changed all the data in the column to Y and N. Now we need to update our table.

In [51]:
%%sql
UPDATE housing_data
SET SoldAsVacant = CASE WHEN SoldAsVacant = "Yes" OR SoldAsVacant = "Y" THEN "Y"
ELSE "N" END;

 * mysql+mysqldb://root:***@localhost/housing_data
56477 rows affected.


[]

The query below confirms if SoldAsVacant has been changed as needed

In [52]:
%%sql
select *
from housing_data
limit 5;

 * mysql+mysqldb://root:***@localhost/housing_data
5 rows affected.


[(0, '105 03 0D 008.00', 'RESIDENTIAL CONDO', datetime.date(2013, 1, 24), Decimal('132000.00'), '20130128-0008725', 'N', '', Decimal('0.000'), '', Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), 0, 0, 0, 0, '1208  3RD AVE S', 'NASHVILLE', '', '', ''),
 (1, '105 11 0 080.00', 'SINGLE FAMILY', datetime.date(2013, 1, 11), Decimal('191500.00'), '20130118-0006337', 'N', 'STINSON, LAURA M.', Decimal('0.170'), 'URBAN SERVICES DISTRICT', Decimal('32000.00'), Decimal('134400.00'), Decimal('168300.00'), 1941, 2, 1, 0, '1802  STEWART PL', 'NASHVILLE', '1802  STEWART PL', ' NASHVILLE', ' TN'),
 (2, '118 03 0 130.00', 'SINGLE FAMILY', datetime.date(2013, 1, 18), Decimal('202000.00'), '20130124-0008033', 'N', 'NUNES, JARED R.', Decimal('0.110'), 'CITY OF BERRY HILL', Decimal('34000.00'), Decimal('157800.00'), Decimal('191800.00'), 2000, 3, 2, 1, '2761  ROSEDALE PL', 'NASHVILLE', '2761  ROSEDALE PL', ' NASHVILLE', ' TN'),
 (3, '119 01 0 479.00', 'SINGLE FAMILY', datetime.date(2013, 1, 18), Decimal('32000.00'), '20130128-0008863', 'N', 'WHITFORD, KAREN', Decimal('0.170'), 'URBAN SERVICES DISTRICT', Decimal('25000.00'), Decimal('243700.00'), Decimal('268700.00'), 1948, 4, 2, 0, '224  PEACHTREE ST', 'NASHVILLE', '224  PEACHTREE ST', ' NASHVILLE', ' TN'),
 (4, '119 05 0 186.00', 'SINGLE FAMILY', datetime.date(2013, 1, 23), Decimal('102000.00'), '20130131-0009929', 'N', 'HENDERSON, JAMES P. & LYNN P.', Decimal('0.340'), 'URBAN SERVICES DISTRICT', Decimal('25000.00'), Decimal('138100.00'), Decimal('164800.00'), 1910, 2, 1, 0, '316  LUTIE ST', 'NASHVILLE', '316  LUTIE ST', ' NASHVILLE', ' TN')]

The SoldAsVacant column is now changed

### v) Deleting irrelevant columns

Deleting columns is not a standard practice. We should always consult before deleting any data from the database.

Assuming we do not need some columns like OwnerAddress, PropertyAddress, and OwnerName because they provide Personal Identifiable Information (PII).

In [54]:
%%sql
ALTER TABLE housing_data
DROP COLUMN OwnerName,
DROP COLUMN OwnerAddress,
DROP COLUMN PropertyAddress;

 * mysql+mysqldb://root:***@localhost/housing_data
0 rows affected.


[]

We need to check if the columns were dropped by printing the top 5 rows of data

In [56]:
%%sql
SELECT * 
FROM housing_data
LIMIT 5;

 * mysql+mysqldb://root:***@localhost/housing_data
5 rows affected.


[(0, '105 03 0D 008.00', 'RESIDENTIAL CONDO', datetime.date(2013, 1, 24), Decimal('132000.00'), '20130128-0008725', 'N', Decimal('0.000'), '', Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), 0, 0, 0, 0, 'NASHVILLE', '', ''),
 (1, '105 11 0 080.00', 'SINGLE FAMILY', datetime.date(2013, 1, 11), Decimal('191500.00'), '20130118-0006337', 'N', Decimal('0.170'), 'URBAN SERVICES DISTRICT', Decimal('32000.00'), Decimal('134400.00'), Decimal('168300.00'), 1941, 2, 1, 0, 'NASHVILLE', ' NASHVILLE', ' TN'),
 (2, '118 03 0 130.00', 'SINGLE FAMILY', datetime.date(2013, 1, 18), Decimal('202000.00'), '20130124-0008033', 'N', Decimal('0.110'), 'CITY OF BERRY HILL', Decimal('34000.00'), Decimal('157800.00'), Decimal('191800.00'), 2000, 3, 2, 1, 'NASHVILLE', ' NASHVILLE', ' TN'),
 (3, '119 01 0 479.00', 'SINGLE FAMILY', datetime.date(2013, 1, 18), Decimal('32000.00'), '20130128-0008863', 'N', Decimal('0.170'), 'URBAN SERVICES DISTRICT', Decimal('25000.00'), Decimal('243700.00'), Decimal('268700.00'), 1948, 4, 2, 0, 'NASHVILLE', ' NASHVILLE', ' TN'),
 (4, '119 05 0 186.00', 'SINGLE FAMILY', datetime.date(2013, 1, 23), Decimal('102000.00'), '20130131-0009929', 'N', Decimal('0.340'), 'URBAN SERVICES DISTRICT', Decimal('25000.00'), Decimal('138100.00'), Decimal('164800.00'), 1910, 2, 1, 0, 'NASHVILLE', ' NASHVILLE', ' TN')]

The Personal Identifiable Information (PII) has been removed from the data.

### vi) Create a SaleMonth column.

We need to create a SaleMonth column from the SaleDate column so that it can be easy to identify which months of the year have high sales.

We shall first create a select statement to extract the moth from date.

In [62]:
%%sql
SELECT SaleDate, EXTRACT(month FROM SaleDate)
FROM housing_data
ORDER BY SaleDate DESC
limit 5;

 * mysql+mysqldb://root:***@localhost/housing_data
5 rows affected.


[(datetime.date(2019, 12, 13), 12),
 (datetime.date(2019, 5, 16), 5),
 (datetime.date(2016, 10, 31), 10),
 (datetime.date(2016, 10, 31), 10),
 (datetime.date(2016, 10, 31), 10)]

In [ ]:
Now we need to alter our table to create a SaleMonth column

In [63]:
%%sql
ALTER TABLE housing_data
ADD SaleMonth INT;

 * mysql+mysqldb://root:***@localhost/housing_data
0 rows affected.


[]

Now we update out table to populate out SaleMonth column

In [64]:
%%sql
UPDATE housing_data
SET SaleMonth = EXTRACT(month FROM SaleDate);

 * mysql+mysqldb://root:***@localhost/housing_data
56477 rows affected.


[]

We now print the top 5 rows to confirm that the SaleMonth column is created

In [65]:
%%sql
SELECT *
FROM housing_data
LIMIT 5;

 * mysql+mysqldb://root:***@localhost/housing_data
5 rows affected.


[(0, '105 03 0D 008.00', 'RESIDENTIAL CONDO', datetime.date(2013, 1, 24), Decimal('132000.00'), '20130128-0008725', 'N', Decimal('0.000'), '', Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), 0, 0, 0, 0, 'NASHVILLE', '', '', 1),
 (1, '105 11 0 080.00', 'SINGLE FAMILY', datetime.date(2013, 1, 11), Decimal('191500.00'), '20130118-0006337', 'N', Decimal('0.170'), 'URBAN SERVICES DISTRICT', Decimal('32000.00'), Decimal('134400.00'), Decimal('168300.00'), 1941, 2, 1, 0, 'NASHVILLE', ' NASHVILLE', ' TN', 1),
 (2, '118 03 0 130.00', 'SINGLE FAMILY', datetime.date(2013, 1, 18), Decimal('202000.00'), '20130124-0008033', 'N', Decimal('0.110'), 'CITY OF BERRY HILL', Decimal('34000.00'), Decimal('157800.00'), Decimal('191800.00'), 2000, 3, 2, 1, 'NASHVILLE', ' NASHVILLE', ' TN', 1),
 (3, '119 01 0 479.00', 'SINGLE FAMILY', datetime.date(2013, 1, 18), Decimal('32000.00'), '20130128-0008863', 'N', Decimal('0.170'), 'URBAN SERVICES DISTRICT', Decimal('25000.00'), Decimal('243700.00'), Decimal('268700.00'), 1948, 4, 2, 0, 'NASHVILLE', ' NASHVILLE', ' TN', 1),
 (4, '119 05 0 186.00', 'SINGLE FAMILY', datetime.date(2013, 1, 23), Decimal('102000.00'), '20130131-0009929', 'N', Decimal('0.340'), 'URBAN SERVICES DISTRICT', Decimal('25000.00'), Decimal('138100.00'), Decimal('164800.00'), 1910, 2, 1, 0, 'NASHVILLE', ' NASHVILLE', ' TN', 1)]

## CONCLUSION

A lot has been done in this project to ensure data completeness, integrity, and relevance. The data types, like dates, are now correctly formatted for better analysis, some missing data points were filled, and new variables have been created. This project uses various MySQL techniques.